extensions/functions/src/main/java/io/crate/window/OffsetValueFunctions.java

```diff
@@ -89,29 +229,36 @@ public Signature boundSignature() {
     public Object execute(int idxInPartition,
                           WindowFrameState currentFrame,
                           List<? extends CollectExpression<Row, ?>> expressions,
-                          Input... args) {
+                          boolean ignoreNulls,
+                          Input[] args) {
         final int offset;
         if (args.length > 1) {
             Object offsetValue = args[1].value();
-            if (offsetValue == null) {
-                return null;
-            } else {
+            if (offsetValue != null) {
                 offset = ((Number) offsetValue).intValue();
+            } else {
+                return null;
             }
         } else {
             offset = 1;
         }
-
-        var lagRowCells = currentFrame
-            .getRowInPartitionAtIndexOrNull(offsetDirection.getTargetIndex(idxInPartition, offset));
-        if (lagRowCells != null) {
-            var lagRow = new RowN(lagRowCells);
-            for (CollectExpression<Row, ?> expression : expressions) {
-                expression.setNextRow(lagRow);
-            }
-
-            return args[0].value();
-        } else {
+        if (offset == 0) {
+            ignoreNulls = false;
+        }
+        // if offset is not constant but varying per iteration, OffsetDirection.cachedNonNullIndex cannot be utilized
+        if (idxInPartition == 0 || (cachedOffset != null && cachedOffset != offset)) {
+            offsetDirection = resolveOffsetDirection(offset);
+        }
```

Comment:
Might be safer to check if the `Input` is a `Literal` to determine if it is a constant. 
Usually it is not a good to check the `Input` for a concrete implementation, but I think in this case it may be justified.

Reply:
@mfussenegger Hello Jordi, I need clarification here. Maybe my comment confused you in the first place. What I wanted to do here was to `resolveOffsetDirection` if this is the first invocation or the offset has changed. My understanding is that no matter the offset is dynamic or from literal, as long as it is the same as previous iteration, the OffsetDirection instance can be reused, otherwise it needs a new one.